Add parallel debugging section (#55)
flux-framework/flux-docs: debugging.rst

.. _debugging:

==============
Debugging Jobs 
==============

Flux supports parallel debuggers such as Rogue Wave Software (RWS)'s

> TotalView parallel debugger (https://totalview.io).

Reviewer: Nitpick. Feel free to ignore:

```suggestion
`TotalView parallel debugger <(https://totalview.io)>`_.
```

  sufficient debug information when it is invoked with no tool control.

Flux provides ``MPIR_partial_attach_ok`` `[1] <https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf>`_
so you can attach TotalView to a subset of parallel processes as well, which
can be handy when you debug a large-scale job. Please refer to
`TotalView user guide`_ on its subset-attach capability.

.. _TotalView user guide: https://docs.roguewave.com/en/totalview/current/html/

----------------------------------------------
Better Handling of Flux's Internal Exec Events
----------------------------------------------

Flux executes a series of subcommands before launching
your parallel program.
Thus, a side effect of the above commands is that TotalView
will stop Flux's own program execution each time
a new subcommand is ``exec(3)``'ed (e.g., when ``flux`` ``exec``'s
``flux-mini``) before your parallel program processes are spawned.
This means you must manually click "Yes" whenever a stop-at-exec
dialogue window is popped up. You do not need to stop
it unless you are a Flux developer!
Instead, you can make TotalView automatically
continue on these events by building the the following

> exec-handling Tcl code into TotalView.
> 
> .. code-block::
> 
>     catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }

Reviewer: Unfortunately, this block is not rendering in readthedocs right now.  I wonder if it is because the language is missing (although the docs say in Sphinx 2.0 it became optional):

```suggestion
exec-handling Tcl code into TotalView.

.. code-block:: tcl

    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }
```

If that doesn't work, it looks like you can get a literal block without the code block directive by ending the previous line with `::` and indenting the actual block [[1]](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#literal-blocks):

```suggestion
exec-handling Tcl code into TotalView::

    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }
```

Author: I tried with Tcl and it didn't work. I will just go with the second suggestion.

.. _debugging:

==============
Debugging Jobs 
==============

> Flux supports parallel debuggers such as Rogue Wave Software (RWS)'s
> TotalView parallel debugger (https://www.roguewave.com).

Reviewer: ReST has support for external links.  The syntax is weird:

```suggestion
Flux supports parallel debuggers such as `Rogue Wave Software (RWS)'s <https://www.roguewave.com>`_
TotalView parallel debugger.
```

Author: Learned a new thing! Will change.

----------------------------------
Parallel Debugging using TotalView
----------------------------------

Debugging your MPI job from the beginning of
parallel program execution:

> .. code-block:: sh
> 
>   totalview --args flux mini run -N 2 -n 2 ./mpi-program

Reviewer: Nit pick.  ReST has a `console` code-block that supports highlighting of the terminal prompt, command, and output.  Since this is just a single line, it doesn't matter all that much.  Mainly an FYI.

```suggestion
.. code-block:: console

  $ totalview --args flux mini run -N 2 -n 2 ./mpi-program
```

Author: Nice! Pretty powerful. Does ReST support most of GitHub mark down verbatim block ? In this case, for example ```console?

Attaching to an already running job:

.. code-block:: sh

  totalview --args flux mini attach <JOBID>

> Note that you should use ``totalview --args flux mini attach`` whether your

Reviewer: ReSt has a note directive, if you want to use that:

```suggestion
.. note::
You should use ``totalview --args flux mini attach`` whether your
```

Author: Cool!

> Note that you should use ``totalview --args flux mini attach`` whether your
> job has been launched via ``flux mini run`` or ``flux mini submit``.
> In other word, Flux currently does not allow TotalView

Reviewer: ```suggestion
Note that you should use ``totalview --args flux mini attach`` when your
job has been launched via ``flux mini run`` or ``flux mini submit``.
In other words, Flux currently does not allow TotalView
```

to attach to your running code if you just attach it
to the ``flux mini run`` process that controls your code.

> Flux provides ``MPIR_partial_attach_ok`` (https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf)

Reviewer: I **think** it is possible to combine the verbatim mode (``) with hyperlinks, but I don't want to suggest something without trying it first.  If it isn't easily doable, maybe something like:

```suggestion
Flux provides ``MPIR_partial_attach_ok`` `[1]<https://www.mpi-forum.org/docs/mpir-specification-10-11-2010.pdf>`_
```

Author: OK. Will try that.

so you can attach TotalView to a subset of parallel processes as well, which
can be handy when you debug a large-scale job. Please refer to

> TotalView user guide on its subset-attach capability: https://docs.roguewave.com/en/totalview/current/html/.

Reviewer: Since the TotalView user guide is referenced multiple times, you can created a named link rather than an anonymous one (I **think** this name is valid anywhere in this file, maybe even across the entire repo).

```suggestion
`TotalView user guide`_ on its subset-attach capability.

.. _TotalView user guide: https://docs.roguewave.com/en/totalview/current/html/
```

Author: Good idea. Just to make sure: here, the external hyperlink is given with the "normal" syntax?

Reviewer: The Sphinx documentation describes it better than I can: https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links

There is basically two ways to do it: with an inline/anonymous link or with a "ref" and "target definition". For this particular case, I recommend trying the latter since this link it used multiple times. You can define the target once, and then reference it multiple times.

----------------------------------------------

Flux executes a series of subcommands before launching
your parallel program.
Thus, a side effect of the above commands is that TotalView
will stop Flux's own program execution each time
a new subcommand is ``exec(3)``'ed (e.g., when ``flux`` ``exec``'s
``flux-mini``) before your parallel program processes are spawned.
This means you must manually click "Yes" whenever a stop-at-exec
dialogue window is popped up. You do not need to stop
it unless you are a Flux developer!
Instead, you can make TotalView automatically
continue on these events by building the the following
exec-handling Tcl code into TotalView.

.. code-block:: sh

    catch {dset TV::exec_handling { {^(flux|lrun|srun|jsrun)(<python[^>]*>(<flux-job>)?|<bash>|<jsrun>|<jswrap>)*$ go}} }

This code should either be added to the site-wide ``.tvdrc`` file
to enable this for all TotalView sessions
or per-user file (e.g., ``tvdrc`` in the current working directory)
to enable this only for the user's own sessions.
Notice that it is designed to support not only Flux but also SLURM's
srun and IBM JSM's jsrun commands. The ``regex`` syntax of

> ``exec_handling`` within TotalView can be found in the TotalView reference
> guide: e.g., https://docs.roguewave.com/en/totalview/current/html/.

Reviewer: Assuming the named reference suggested above is created:

```suggestion
``exec_handling`` within TotalView can be found in the `TotalView user
guide`_.
```
